metavalidator/any.go

```diff
@@ -0,0 +1,72 @@
+package metavalidator
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+)
+
+var _ tfsdk.AttributeValidator = anyValidator{}
+
+// anyValidator validates that value validates against at least one of the value validators.
+type anyValidator struct {
+	valueValidators []tfsdk.AttributeValidator
+}
+
+// Description describes the validation in plain text formatting.
+func (v anyValidator) Description(ctx context.Context) string {
+	var descriptions []string
+	for _, validator := range v.valueValidators {
+		descriptions = append(descriptions, validator.Description(ctx))
+	}
+
+	return fmt.Sprintf("Value must satisfy at least one of the validations: %s", strings.Join(descriptions, " + "))
+}
+
+// MarkdownDescription describes the validation in Markdown formatting.
+func (v anyValidator) MarkdownDescription(ctx context.Context) string {
+	return v.Description(ctx)
+}
+
+// Validate performs the validation.
+// If the diagnostics returned from the validator does not contain an error
+// we return all accumulated warning diagnostics.
+func (v anyValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, resp *tfsdk.ValidateAttributeResponse) {
+	for _, validator := range v.valueValidators {
+		validatorResp := &tfsdk.ValidateAttributeResponse{
+			Diagnostics: diag.Diagnostics{},
+		}
+
+		validator.Validate(ctx, req, validatorResp)
+
+		if !validatorResp.Diagnostics.HasError() {
+			diagWarnings := diag.Diagnostics{}
+
+			for _, d := range resp.Diagnostics {
```

Comment:
This change will return all of the accumulated `warning` diagnostics from previously failed validations along with any `warning` from the validator that passes as [discussed](https://hashicorp.slack.com/archives/C030RBBA63U/p1656348677761799?thread_ts=1656347176.051779&cid=C030RBBA63U) - _"For the Warning for Any, I think it’s non desirable to swallow those"_.

Reply:
Have created a separate AnyWithAllWarningsValidator but more than happy to change the name if something else would suit better.